### DISCUSSION TITLE
UWSGI, Nginx, Flask, Socket.IO troubles

### DISCUSSION THREAD

**Discussion by jimploof** [miguelgrinberg/Flask-SocketIO#2104] [Q&A]
I have an existing single page resume application that working just fine in production using flask, python, uwsgi, and nginx. I wanted to implement chat functionality into the application, but have had nothing but trouble. 

In an effort to simplify things I have created a UWSGI emperor, and vassals, one is intended to serve the main application, which it is doing, The other was to deal with the flask socket.io functionality, which depending on the configuration, i have tried many, is either returning 502 or 504 errors. 

I have read over suggested documents, and like i said tried dozens and dozens of different configurations with nginx/uwsgi/app.py.

Current /etc/nginx/sites-available/nginx.conf 
`# HTTP to HTTPS redirect for all domains
server {
    listen 80;
    server_name jimploof.com www.jimploof.com jamesploof.com www.jamesploof.com portfolio.jimploof.com;

    # Redirect all HTTP traffic to HTTPS for jimploof.com
    return 301 https://jamesploof.com$request_uri;
}

server {
    listen 443 ssl;
    server_name jimploof.com www.jimploof.com portfolio.jimploof.com;

    # SSL certificates specific to jimploof.com
    ssl_certificate /etc/letsencrypt/live/jimploof.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/jimploof.com/privkey.pem;
    include /etc/letsencrypt/options-ssl-nginx.conf;
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;

    # Redirect to jamesploof.com
    return 301 https://jamesploof.com$request_uri;
}

server {
    listen 443 ssl;
    server_name jamesploof.com;

    # SSL certificates specific to jamesploof.com
    ssl_certificate /etc/letsencrypt/live/jamesploof.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/jamesploof.com/privkey.pem;
    include /etc/letsencrypt/options-ssl-nginx.conf;
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;

    # uWSGI serving the Flask application
    location / {
        include uwsgi_params;
        uwsgi_pass unix:/var/www/resume/resume.sock;

        # Pass GeoIP2 data as headers to Flask
        uwsgi_param X-GeoIP2-Country-Code $geoip2_data_country_code;
        uwsgi_param X-GeoIP2-Country-Name $geoip2_data_country_name;
        uwsgi_param X-GeoIP2-City-Name $geoip2_data_city_name;
        uwsgi_param X-GeoIP2-Latitude $geoip2_data_latitude;
        uwsgi_param X-GeoIP2-Longitude $geoip2_data_longitude;
    }

location /socket.io/ {
    proxy_pass http://unix:/var/www/socketio/socketio.sock;
    #proxy_pass http://localhost:8888/socket.io/;
    proxy_http_version 1.1;
    proxy_buffering off;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-Host $host;
    proxy_set_header X-Forwarded-Port $server_port;

    proxy_read_timeout 300s;
    proxy_send_timeout 300s;
    proxy_connect_timeout 300s;

    error_log /var/log/nginx/socketio-error.log debug;
    access_log /var/log/nginx/socketio-access.log combined;

    add_header X-Debug-Message "Proxied through Socket.IO location" always;
}

location /socket/ {
    include uwsgi_params;
    uwsgi_pass unix:/var/www/socketio/socketio.sock;
    uwsgi_param SCRIPT_NAME /socket;
    uwsgi_modifier1 30;

    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "Upgrade";
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
}

    location /go/ {
        proxy_pass http://localhost:8080/;  
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        # WebSocket support for Go backend
        proxy_read_timeout 300s;
        proxy_send_timeout 300s;
    }

    add_header X-Frame-Options "SAMEORIGIN";
    add_header X-XSS-Protection "1; mode=block";
    add_header X-Content-Type-Options "nosniff";
    add_header Referrer-Policy "strict-origin-when-cross-origin";
}

error_log /var/log/nginx/websocket_error.log debug;
`

current socket uwsgi vassal 
`[uwsgi]
module = app:app
master = true
processes = 1
socket = /var/www/socketio/socketio.sock
chmod-socket = 666
vacuum = true
die-on-term = true
enable-threads = true
single-interpreter = true
chdir = /var/www/socketio
virtualenv = /var/www/socketio/venv
plugins = python3
uid = www-data
gid = www-data
buffer-size = 65535
log-master = true
logto = /var/www/socketio/uwsgi.log
http-websockets = true
harakiri = 120
post-buffering = 65535
thunder-lock = true
max-requests = 2000
worker-reload-mercy = 60
reload-mercy = 60
`

and finally app.py
`# /var/www/socketio/app.py
import eventlet
eventlet.monkey_patch()

import os
import logging
import sys
from flask import Flask, render_template, request, jsonify
from flask_socketio import SocketIO

# Enhanced logging setup
logging.basicConfig(
    filename='/var/www/socketio/socketio.log',
    level=logging.DEBUG,
    format='%(asctime)s %(levelname)s: %(message)s [in %(pathname)s:%(lineno)d]'
)

# Also log to stderr for uWSGI
handler = logging.StreamHandler(sys.stderr)
handler.setLevel(logging.DEBUG)
logging.getLogger().addHandler(handler)

app = Flask(__name__)
app.config['SECRET_KEY'] = 'dev'
app.config['DEBUG'] = True

socketio = SocketIO(
    app,
    async_mode='eventlet',
    logger=True,
    engineio_logger=True,
    cors_allowed_origins="*"
)

@app.route('/socket.io/test')
def test_client():
    app.logger.info(f'Test client accessed from {request.remote_addr}')
    try:
        return render_template('index.html')
    except Exception as e:
        app.logger.error(f'Error serving test client: {str(e)}')
        return str(e), 500

# Add a health check route
@app.route('/socket.io/health')
def health_check():
    return jsonify({'status': 'healthy'})

@app.errorhandler(Exception)
def handle_error(e):
    app.logger.error(f'Unhandled error: {str(e)}')
    return str(e), 500

@socketio.on_error()
def error_handler(e):
    app.logger.error(f'SocketIO error: {str(e)}')

@socketio.on('connect')
def handle_connect():
    app.logger.info(f'Client connected: {request.sid} from {request.remote_addr}')

@socketio.on('disconnect')
def handle_disconnect():
    app.logger.info(f'Client disconnected: {request.sid}')

@socketio.on('message')
def handle_message(data):
    app.logger.info(f'Received message from {request.sid}: {data}')
    try:
        socketio.emit('response', {'data': f'Server received: {data}'})
    except Exception as e:
        app.logger.error(f'Error sending response: {str(e)}')

if __name__ == '__main__':
    socketio.run(app)
`

if i can figure this out ill definitely be giving back to community and writing a tutorial, thanks in advance!

**Comment by jimploof** [comment]
I updated my nginx to use an ip address:port, and then updated my uwsgi to use http:port, and now I am getting a new error about "The client is using an unsupported version of the Socket.IO or Engine.IO protocols"

`bidict             0.23.1
blinker            1.8.2
certifi            2024.8.30
charset-normalizer 3.3.2
click              8.1.7
dnspython          2.7.0
eventlet           0.37.0
Flask              3.0.3
Flask-Cors         5.0.0
Flask-JWT-Extended 4.6.0
Flask-Mail         0.10.0
Flask-SocketIO     5.1.2
gevent             24.10.2
gevent-websocket   0.10.1
greenlet           3.1.1
h11                0.14.0
idna               3.8
itsdangerous       2.2.0
Jinja2             3.1.4
MarkupSafe         2.1.5
pika               1.3.2
pip                24.2
PyJWT              2.9.0
python-dotenv      1.0.1
python-engineio    4.3.2
python-socketio    5.6.0
redis              5.0.8
requests           2.32.3
setuptools         65.5.0
simple-websocket   1.1.0
six                1.16.0
urllib3            2.2.3
uWSGI              2.0.26
websocket-client   1.8.0
Werkzeug           3.0.4
wheel              0.44.0
wsproto            1.2.0
zope.event         5.0
zope.interface     7.1.0
`

**Reply by miguelgrinberg** [reply]
The **client** is using an unsupported version. See the version compatibility table in the documentation and use versions of the client and server that are compatible with each other.

**Reply by jimploof** [reply]
Hi Miguel, I did view that documentation, and I am using Flask 3.0.3, python-engineio    4.3.2, python-socketio    5.6.0, Flask-SocketIO     5.1.2...

//cdnjs.cloudflare.com/ajax/libs/socket.io/4.6.0/socket.io.js

Which I believe all should be compatible according to documentation?

**Reply by miguelgrinberg** [reply]
If your are 100% sure the client and the server are compatible, then something that sits in between must be modifying the requests in a way that makes the server think they aren't from a compatible version. The client announces the version in the `EIO` query string argument. You will need to debug your requests to determine where is the query string getting changed.

**Comment by jimploof** [comment]
The only thing I am seeing in logs is the following 

> 2024-10-22 21:55:10,513 ERROR: The client is using an unsupported version of the Socket.IO or Engine.IO protocols (further occurrences of this error will be logged with level INFO) [in /var/www/socketio/venv/lib/python3.11/site-packages/engineio/base_server.py:334]
2024-10-22 21:57:00,984 INFO: The client is using an unsupported version of the Socket.IO or Engine.IO protocols [in /var/www/socketio/venv/lib/python3.11/site-packages/engineio/base_server.py:338]

> 2024/10/22 21:55:10 [error] 564275#0: *12 connect() failed (111: Connection refused) while connecting to upstream, client: 165.225.11.61, server: jamesploof.com, request: "GET /socket.io/test HTTP/1.1", upstream: "http://[::1]:8888/socket.io/test", host: "jamesploof.com"
2024/10/22 21:55:10 [warn] 564275#0: *12 upstream server temporarily disabled while connecting to upstream, client: 165.225.11.61, server: jamesploof.com, request: "GET /socket.io/test HTTP/1.1", upstream: "http://[::1]:8888/socket.io/test", host: "jamesploof.com"

> The client is using an unsupported version of the Socket.IO or Engine.IO protocols
[pid: 564295|app: 0|req: 2/2] 127.0.0.1 () {50 vars in 875 bytes} [Tue Oct 22 21:57:00 2024] GET /socket.io/test => generated 84 bytes in 0 msecs (HTTP/1.1 400) 1 headers in 54 bytes (1 switches on core 0)

> wscat -c ws://127.0.0.1:8888/socket.io/test/?EIO=4&transport=websocket
> [1] 568788
> (venv) guht@jimploof-ubuntu-2vcpu-4gb:/var/www/socketio$ fg %1
> wscat -c ws://127.0.0.1:8888/socket.io/test/?EIO=4
> error: Unexpected server response: 200


application log, nginx, and uwsgi logs.... 

not sure how else to troubleshoot, not seeing the typical EIO4 in any requests, im assuming thats determined automatically somehow.

**Reply by miguelgrinberg** [reply]
I'm sorry I don't know how I can help. You are clearly not using a normal Socket.IO client since none of the requests you have in your logs follow the pattern of what the JS Socket.IO client sends. You may want to try a simpler set up using the official Socket.IO client. Testing Socket.IO with wscat is also not appropriate, the Socket.IO protocol is complex and requires a smart client that understands the protocol.

**Reply by jimploof** [reply]
I appreciate you taking the time to read and respond! truly you are a great asset to the community and.
